### PR TITLE
Tarantool jcache

### PR DESCRIPTION
**Pull request by rave4life**

### CODE REVIEW

**Comment by rave4life** [comment]
Alexander, thank you for your detailed review.

First of all let's talk about architecture of Cache Implementation. We should implement JSR107 API (including Cache Event Listeners support) -  this way was chosen for some reasons:
- JSR107 (JCache) can be used with Hibernate ORM. [JCacheRegion](http://docs.jboss.org/hibernate/orm/5.3/javadocs/org/hibernate/cache/jcache/JCacheRegion.html) is used to plug Tarantool JCache with Hibernate. And finally we get so called Hibernate+JCache solution.
- Hibernate provides concurrency. As you mentioned (in Note 1) JSR107 has concurrency issues. It's true but Hibernate helps developers to solve this by providing CacheConcurrencyStrategy.READ_WRITE concurrency strategy. Now it is already supported in Tarantool JCache, but we should perform full consistency and competitive access test and performance test with workload to make final conclusion about this way.
- Hibernate has well documented annotations which can be used in Java applications
- Hibernate implements 2 levels of caching: L1 and L2. L1 is local in-memory cache and is related to current persistence context. L2 is related to SessionFactory. Tarantool JCache is L2 cache. So question is: could this feature be used as two-tier caching? Or it is not exactly the same? EhCache uses multi-tiered caching apart from Hibernate's ones.
- Hibernate is simple to plug/unplug Cache Providers (via Service Provider Interfaces). In case of unsuccessful results Tarantool JCache can be easily replaced by another cache providers (Hazelcast, EhCache, ...) without interventions to Java code. Also EhCache can be used with Hibernate without JCacheRegion, but there are no easy ways to implement all the Hibernate's features w/o using JCache.

But if we talk about Write-Behind strategy (mentioned in your Note 2), it wasn't elaborated and tested yet. Perhaps CacheLoader/CacheWriter should be used there, but also as a part of JSR-107. CacheLoader/CacheWriter interfaces are used within JCache implementation to load/write data from/to database, this interfaces cannot be plugged/unplugged in declarative form. Fix me if I'm wrong. We could discuss this.

A few wolds about [JCacheRegionFactory](http://docs.jboss.org/hibernate/orm/5.3/javadocs/org/hibernate/cache/jcache/JCacheRegionFactory.html) . CacheLoader and CacheWriter are not used (always null). Hibernate uses own way to implement Write-Through/Read-Through mechanism. Also default configuration is created and passed to TarantoolCache (with expiration/eviction parameters such a ttl, memtx_memory and so on). We should use declarative configuration of cache via xml to customize Tarantool JCache.
So my question is how to make xml schema for Tarantool JCache? Perhaps xsd schema is already exists in Tarantool community or in enterprise solutions? EhCache/Hazelcast uses own schema.

Let's talk about your review notes.
> Copyrights & license. It is not clear under what circumstances the code were shared.

Removed this reference.

> Change of package version to 1.8.jcache-SNAPSHOT.

Yes, agree with you. My idea was to get your opinion about this way. So README.md would be also integrated with existing one.

> Windows line endings.

Ok, will be replaced.

> Maybe JSRMsgPackLite methods can be just merged into MsgPackLite?

There is only one issue prevents me from merging JSRMsgPackLite into MsgPackLite. JSRMsgPackLite uses MP_BIN32 to identify binary Java Objects. I think it is not good idea to make assumption that all "MP_BIN32" fields retrieved from Tarantool's tuple are Java Objects. Should we introduce new type id in MsgPackLite for Java Objects (perhaps MP_EXT)? I'll try and test it for amount versions of a Tarantool's core.

> As far as I know star imports are discouraged.

Fixed.

> Why was changed to Object in CachedValue?

Already fixed.

> Why add/remove listeners code was removed in TarantoolCache?

Listeners will be supported soon.

> Should we drop on close? Maybe it is better to sync on instanciate?

Yes, we should't drop/truncate spaces on close. This feature was left to make TCK tests succeeded. Drop becomes disabled after eviction strategy would be implemented.

> TarantoolCache.{containsKey, put}: why locking was removed?

Locker hasn't implemented in a proper way, so it was removed. Hibernate's lock strategy used instead. If it wouldn't be enough to keep consistency a lock strategy with Tarantool's triggers would be used.

> loadAll uses lambda function, so Java 8+ only; don't see a reason for that.

Yes, the decision is to make a simpler construction.

> Are we need to plan support of 2-tier topology ... support in the future to achieve
competitive access speed?

We need to discuss this, it is important. As mentioned above Hibernate already has L1 and L2 caching levels - but maybe it's not enough now to achieve competitive access speed.

> Expired entries are not subject to remove in background to free memory?

Tarantool's expirationd daemon is candidate to take this care, but it does this routine asynchronously. So we need to discuss how to implement this offload in a proper way and keep consistency. Also we can run additional java thread within JCache. Or if we talk about triggers or stored functions should we take into account access speed?

> Eviction strategy: How consumed memory is limited? In sync with
memtx_memory option?

Yes, it would be great to sync with memtx_memory option. But this option is related to all spaces, not to just one. So Tarantool JCache has to decide from which space tuple should be evicted (current space where Insert/Upsert fails or most overfull space). Perhaps there is a good idea to configure strategy via xml. Also extended index should be created to find least recently used tuple to be evicted as fast as possible.

> Should we think of vshard on backend in the future?

Yes, we should.

> Raw idea: CacheValue as fabric / builder that encapsulates tarantool
changes?

Already thought of this trend, but it is unclear yet how to do it in a maximal convenient way.

> Got much of errors in tarantool log ... 

Please give me some recommendations about tests, especially what environment should be used. Already fixed some bugs related to key and value field type, so JCache's key is able to stored as Java Object. Currently no errors were found in log.

Finally please tell me should I commit to this branch or it's appropriate to make new branch for further commit(s) and pull request?

**Comment by Totktonada** [comment]
I appreciate your detailed answer too :)

Let's start the next iteration...

First, let me apologize for being verbose in wording and sometimes too formal.
It is how gears work in my brain...

> First of all let's talk about architecture of Cache Implementation. We should implement JSR107 API (including Cache Event Listeners support) - this way was chosen for some reasons:
> ...

I look how other storage / database products exposes its API and, yep, it is
JSR107 implementations. Implementing the standard (instead of 'plugin' to
existing product(s)) opens a way more freedom in design choices. The example
we seen already: it allows waive of on-heap cache.

So now I agree we should implement full JSR107.

Hovewer I have two matters of concern here.

1.
    The API desined in the way that obligates a Cache implementation to do
    non-caching work (loaders/writers, listeners, processors, statistics, ...).
    Many cache providers (I choose some random ones and look at Ehcache,
    Infinispan, Caffeine) have its own cache API that does not involves
    unrelated concepts and have JSR107 cache implementation as a proxy or
    adaptor (Caffeine's name CacheProxy looks most meaningful for me comparing
    ${ProductName}Cache in others).

    It seems there are persons aside me who concer about such things: I found
    related [question][1] (w/o positive resolution, though). Personally I
    think it would be good if the API would be designed in an another way to
    separate unrelated concepts. Cache concepts would live in its own package
    and cache observers concepts would be in another package, separate
    implementation would be possible. But the API is standartized, so our
    option are the following:

    * Separate JSR107 stuff and tarantool cache stuff per packages within one
      project.
    * Provide abstract JSR107 Cache (w/o caching, just stub with listeners
      logic and so on) as separate package and use it as dependency.

    The second approach is (a bit?) more complex, because introduces a package
    with intention to be general and reusable. Hovewer it has its pros: code
    that are not related to tarantool will be otside of the connector, it can
    be reused and even standartized somehow.

    I complain about that, because now it is hard to concentrate on some
    certain topic (statistics, tarantool protocol/connector usage, heap
    consumption) to try to verify code correctness.

    Fix me if things does not feel so for you.

2. The reference implementation is not mature. We need to choose another one
   as base or give commitment we'll investigate possible breaking scenarious,
   fix and test it appropriatelly. I think choice of base implementation does
   not matter much (because it'll just abstract jsr107 cache) if it is one
   that has significant production usage. I don't push you to rewrite your
   code to, say, ehcache base, but I think you should consider risks and
   amount of work to investigate possible RI problems. 'Write it twice' cost
   is not so much comparing support and further develop costs across some
   reasonable time.

One more thing I have to say here (just for the record): after some
discussions within our team I have formulated our goal: distributed, scalable,
fault-tolerant cache of Java objects.

Re Hibernate usage: thanks for the clarification. I'm not the Java guy, so
details are helpful. I think the way you use the product should be formulated
(in README?) as recommendations for users and/or maybe integrated (in pom.xml
as comments?). But maybe I fired this too early, so I just leave note here and
will go ahead.

> So my question is how to make xml schema for Tarantool JCache? Perhaps xsd schema is already exists in Tarantool community or in enterprise solutions? EhCache/Hazelcast uses own schema.

I don't hear about such schemas. If you are about configuring tarantool
instance(s) (box.cfg options) from Java: don't sure it is right thing to do
(but it is interesting feature, though), but it is possible to just map
available box.cfg options into xml schema.

> > Maybe JSRMsgPackLite methods can be just merged into MsgPackLite?
>
> There is only one issue prevents me from merging JSRMsgPackLite into MsgPackLite. JSRMsgPackLite uses MP_BIN32 to identify binary Java Objects. I think it is not good idea to make assumption that all "MP_BIN32" fields retrieved from Tarantool's tuple are Java Objects. Should we introduce new type id in MsgPackLite for Java Objects (perhaps MP_EXT)? I'll try and test it for amount versions of a Tarantool's core.

Let me rephrase the problem. We have 'old' encoding, which maps primitive (and
some object) types into msgpack in the way allows tarantool (or another
tarantool connector) understand it as its native type. Now we introduce the
new way to encode an object into msgpack, which allows to encode any object
marked as Serializable, but lacks of ability to interpret resulting blob
outside of Java. We want to leave old encoding when possible to make data
externally interpretable when possible.

So, we have two encodings, whose values can interleave. And so we should
introduce the new column to specify encoding was used. Encodings can be named
as MSGPACK_TARANTOOL and MSGPACK_JAVA_OBJECT (I just propose to make things
clear, feel free to use any name you think more appropriate).

> Yes, we should't drop/truncate spaces on close. This feature was left to make TCK tests succeeded. Drop becomes disabled after eviction strategy would be implemented.

So the behaviour will be changed? Please, mark it with FIXME in the code or
track in some other way (e.g. FIXME in the commit message if it is going to be
changed within the patchset or issue if don't). Don't let things being broken
w/o some 'warning triangle'.

> > Are we need to plan support of 2-tier topology ... support in the future to achieve
competitive access speed?
>
> We need to discuss this, it is important. As mentioned above Hibernate already has L1 and L2 caching levels - but maybe it's not enough now to achieve competitive access speed.

There is opinion in our team that we don't need on-heap cache and tarantool
can be L1 itself. I don't sure, but I think one-tier cache is okay for now
(especially when it is possible to wrap it into Hibernate to get 2-tier
caching).

> > Expired entries are not subject to remove in background to free memory?
>
> Tarantool's expirationd daemon is candidate to take this care, but it does this routine asynchronously. So we need to discuss how to implement this offload in a proper way and keep consistency. Also we can run additional java thread within JCache. Or if we talk about triggers or stored functions should we take into account access speed?

Now the java cache implementation has two interesting properties:

* It does not reinvent toolset for managing tarantool instance(s) and let
  admin / operation team manage tarantools as they wish using native
  mechanisms (application and caching service can be even provided by
  different teams!).
  - Hovewer some non-inturuisive way to configure tarantool from an
    application maybe will be convenient and maybe will become necessary.
* It does not require tarantool to have any script underneath, just
  box.cfg{listen=...} is enough. (But it is possible to implement some extra
  features.)

Expiration / eviction with expirationd would broke the second property and
obligate admin / operation team to have some specific script on tarantool side
in sync with java part. I think the property (that we don't have this as the
requirement) is tasty, because sinks down entry level for new users and
simplifies deployment / administrative tasks.

I propose to postpone this question for now.

> > Eviction strategy: How consumed memory is limited? In sync with
> > memtx_memory option?
>
> Yes, it would be great to sync with memtx_memory option. But this option is related to all spaces, not to just one. So Tarantool JCache has to decide from which space tuple should be evicted (current space where Insert/Upsert fails or most overfull space). Perhaps there is a good idea to configure strategy via xml. Also extended index should be created to find least recently used tuple to be evicted as fast as possible.

It is related to the two properties above, so I would postpone it too. It is
not clear what we should do, at least for me.

> > Should we think of vshard on backend in the future?

> Yes, we should.

This is +1 to my proposal about separating of JSR107 adaptor and cache
implementations. We'll have tarantool cache and vshard cache in the future,
but hopefully one JSR107 adaptor.

> > Raw idea: CacheValue as fabric / builder that encapsulates tarantool
> > changes?
>
> Already thought of this trend, but it is unclear yet how to do it in a maximal convenient way.

My primary objection was against mixing of JSR107 'boilerplate' code and the
code doing caching within Cache and CacheManager implementations. I was in
thoughts how to handle that and now I think that CacheValue should be part of
JSR107 adaptor, but real caching code should operate on keys and values.

I don't sure about it to be honest, because I don't dive enough into the
problem (I should of course); some things unclear for me:

* How CacheEntry is used (just in processors and so on?)?
* Whether it is easy to 'forget' CacheEntry (feed it to gc) or avoid its
  creation at all (when no processors?) to narrow memory usage.
  
...but separate JSR107 adaptor (with CacheValue just within its scope) is what
I see reasonable on my current level of understanding.

> > Got much of errors in tarantool log like so:
> >
> > > iproto xrow.c:99 E> ER_INVALID_MSGPACK: Invalid MsgPack - packet header
>
> Please give me some recommendations about tests, especially what environment should be used. Already fixed some bugs related to key and value field type, so JCache's key is able to stored as Java Object. Currently no errors were found in log.

I just run `tarantool src/test/instance.lua` (I have tarantool installed
system-wide, the system in Gentoo Linux) and run `mvn verify`. It is just
because it is most simple way for me and is not some kind of recommendation.
Maybe I do something wrong and got these error because of that. I checked
spaces and they are empty after tests, but the cache is truncated when closed,
so it is not an indicator of a problem. I don't investigate it.

Anyway, I'll check it better on some next more deep review.

> Finally please tell me should I commit to this branch or it's appropriate to make new branch for further commit(s) and pull request?

Push (and even force-push) to a feature branch is okay. I don't see reason to
create a new branch and open the new PR. Usually a new version of the patchset
just force-pushed to the feature branch after a review iteration. But feel
free to create separate branch if you want to keep this one for some reason.

----

Can you please formalize backlog (tasks to be resolved in the future) somehow?
Maybe just in a commit message now or (if it is already clear enough) in
issues.

These things are in my focus now:

* Decide about base implementation.
* Separate caching code and JSR107 stuff.
* Implement listeners (if we'll continue with RI as base).
* Resolve truncation at close.
* Investigate concurrent issues, test, fix (at least if we'll use RI).
* Document usage within Hibernate: quickstart, strategies, tiers, locking.
* Implement JSR107 annotations (they are useful I think).
* Support caching in vshard.
* Decide how tarantool instance(s) should be configured (from app / natively
  on tarantool / both).
* Decide how to implement expiration and eviction (from app / from script).

Some of them should be part of MVP (minimal valuable product), some can be
postponed to future milestones.

Please, expend this list if I miss something.

[1]: https://groups.google.com/forum/#!topic/jsr107/T3gV15P9Pys

**Comment by Totktonada** [comment]
Decided to move this work to https://github.com/tarantool/tarantool-jcache
